### PR TITLE
Add support for "debounce" and "failsafe" param

### LINKED ISSUE #27: Option to change debounce wait value and turn off failsafe
Would you consider adding options to change debounce wait value and turning off failsafe check?

Since they apply globally to whole in-viewport module, they can’t be added to each instance, so there should be option to change exported values from module, e.g.

```js
const inviewport = require('in-viewport');
inviewport.debounceWait = 300;
inviewport.failsafe = false;
```

Another option (and one which will cover case of individual listeners), is to modify `instances` array to contain properties `container`, `debounceWait` and `failsafe`, and use all 3 properties for checking if instance already exists. If it does, use it, otherwise create new instance with listeners bound to values that user provided.

What do you think? I can create PR if this is something you think is nice to have.

### PR DESCRIPTION
**Pull request by niksy**
Closes #27.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -98,6 +98,42 @@ function visible() {
...
+function visible() {
+  alert('myfancyDiv is visible in the `customContainer`!');
+}
+```
+
+### Failsafe check
+
+By default, failsafe check with `setInterval` is performed every 150ms.
+
+Usecase: a hidden parent containing eleements; when the parent becomes visible, we have no event that the children became visible.
```
Comment (tzi) [inline README.md]: s/eleements/elements :koala:

File: README.md [hunk 2]
```diff
@@ -98,6 +98,42 @@ function visible() {
 
 When your element is near `300px` of the viewport, you get your callback / true result.
 
+### Specifying debounce wait value
+
+By default, debounce wait value is 15ms.
```
Comment (tzi) [inline README.md]: Could you explicit which debounce we are talking about?

**Comment by tzi** [comment]
Sorry, for the late comment. I forgot to valid this comment in my first review. :disappointed:

**Comment by tzi** [comment]
Thanks a lot for your work here.

This PR is OK for me.

We will fix / add the tests before merging it.
But I will not have the time in the next two days, sorry for the delay.

**Comment by niksy** [comment]
@tzi no problem!

File: README.md [hunk 3]
```diff
@@ -98,6 +98,42 @@ function visible() {
...
+```js
+var inViewport = require('in-viewport');
+var elem = document.getElementById('myFancyDiv');
+
+inViewport(elem, { debounce: 300 }, visible);
+
+function visible() {
+  alert('myfancyDiv is visible in the `customContainer`!');
+}
+```
```
Comment (vvo) [inline README.md]: We need to explain to the user what are the places changing the debounce is smart. What is the usecase of modifying the default debounce?

File: README.md [hunk 4]
```diff
@@ -98,6 +98,42 @@ function visible() {
...
+
+### Failsafe check
+
+By default, inViewport does a failsafe to handle display manipulation that does not throw an event. It works with a `setInterval` performed every 150ms.
```
Comment (vvo) [inline README.md]: I am lost in my own library, what's the code handling that and what's its used for exactly?

File: README.md [hunk 5]
```diff
@@ -98,6 +98,42 @@ function visible() {
...
+
+Usecase: a hidden parent containing elements; when the parent becomes visible, we have no event that the children became visible.
+
+If you don’t want to run that check constantly, turn it off with `failsafe` param.
```
Comment (vvo) [inline README.md]: Why would you not want to have this failsafe? Again we need to guide the user into the best decision by providing him pro and cons, context.

Comment (niksy) [inline README.md]: `setInterval` is called every 150ms which is in most cases unecessary. It is failsafe after all and it covers edge case from what I can see. Maybe it should be disabled by default and left to developer to handle it by themselves? Same thing is with MutationObservers—it is handled in browsers which support them, but for all others, developers are instructed to polyfill support.

File: README.md (hunk 4 above)
Comment (niksy) [inline README.md]: https://github.com/vvo/in-viewport/blob/master/in-viewport.js#L94-L98

File: README.md (hunk 3 above)
Comment (niksy) [inline README.md]: Currently, scroll/resize events are called every 15ms which isn’t that different than 0ms. If we give users option to change it, they can set it to some sensible value like 300ms. Good explanation of usecase can be see on [CSS-Tricks article](https://css-tricks.com/the-difference-between-throttling-and-debouncing/#article-header-id-2).

Comment (vvo) [inline README.md]: 👍

Comment (vvo) [inline README.md]: Can you say that inside the readme then? So that the parameter has some use case explanation

File: README.md (hunk 5 above)
Comment (vvo) [inline README.md]: > Same thing is with MutationObservers—it is handled in browsers which support them, but for all others

Quick note, now MutationObservers are basically available everywhere starting at IE11.

Comment (vvo) [inline README.md]: Before we add a new parameter, we need to explain to the user when he could benefit from it. Thus we need a good explanation of when would using failsafe makes sense. The current explanation about disabling some timers is not sufficient to me.

Comment (vvo) [inline README.md]: "If you don’t want to run that check constantly"

As a user I will ask myself: "Why?". There's no precise answer on facts here so it raises [FUD](https://en.wikipedia.org/wiki/Fear,_uncertainty_and_doubt)

File: README.md (hunk 3 above)
Comment (niksy) [inline README.md]: Sure!

File: README.md (hunk 5 above)
Comment (niksy) [inline README.md]: Yes, you’re right.

Maybe something along the lines: "If you handle cases like usecase above by yourself in different part of your codebase (e.g. you have callback which is active when parent becomse visible), you can turn it off with `failsafe` param."

Comment (vvo) [inline README.md]: > cases like usecase

is not ready to read but something similar should work, try it and update PR

Comment (niksy) [inline README.md]: I’ve updated the description.

File: README.md (hunk 3 above)
Comment (niksy) [inline README.md]: Done!

**Comment by niksy** [comment]
Is there anything more I can help?

**Comment by vvo** [comment]
We are good, thanks a lot

**Comment by niksy** [comment]
@vvo I was asking because there aren’t any tests since I was having trouble creating them (mentioned in https://github.com/vvo/in-viewport/issues/27#issuecomment-294494321) and right now when I’m checking latest published version, `failsafe` parameter is ignored since it has loose check so when you set it to `false` it won’t get applied.

This was error on my part, unfortunately :( Do you also have problems with executing tests locally?

**Comment by vvo** [comment]
Locally tests are ok, use npm run dev to add more.

This library code and stack is VERY old, if you would like to takeover and change a lot of things on how tests are written, code is written etc.. Please do, I am looking for active maintainers :)

**Pull request merged**